### DISCUSSION TITLE
Serious serialization problem in v2, super easy in v1: globally format datetimes in v2

### DISCUSSION THREAD

**Discussion by etienne-monier** [pydantic/pydantic#7405] [Question] [answered]
Hi,

It seems I can't figure out a serialization problem in v2 which is simple in v1.

### The problem

I use FastAPI models with human-readable datetime. Most of them have nested models with a lots of such datetime attributes.

Sometimes, the users want to get a report with precise datetime.

In such case, instead of parsing the database entry with the human-readable model, I use a model with precise datetimes.

### How to do that in v1

In v1, I simply overwrite the `json_encoders` model config. Something like this.

```python
class ParentModel(BaseModel):
    """A parent model with human readable datetimes.

    All its properties and child model properties should have %Y-%m-%d %H:%M format."""

    c: ChildModel
    d: HumanDatetime
    
    class Config:
        json_encoders={
            datetime: lambda v: v.strftime(r"%Y-%m-%d %H:%M"),
        }


class ParentModel(BaseModel):
    """A ParentModel with high precision datetimes.

    All its properties and child model properties should have ISO format with nanoseconds."""

    class Config:
        json_encoders={
            datetime: lambda v: v.isoformat(),
        }
```

### The v2 proposal

I don't want to use the [`field_serializer`](https://docs.pydantic.dev/latest/api/functional_serializers/#pydantic.functional_serializers.field_serializer) as many of the sub-models have datetime attributes.

I tough to use the `PlainSerializer` as [here](https://docs.pydantic.dev/latest/usage/serialization/#custom-serializers), which leads to the following code.

```python
from datetime import datetime
from typing import Annotated

from pydantic import BaseModel, ConfigDict
from pydantic.functional_serializers import PlainSerializer

HumanDatetime = Annotated[
    datetime,
    PlainSerializer(lambda x: x.strftime(r"%Y-%m-%d %H:%M"), return_type=str, when_used="json"),
]
"""A humman readable datetime annotation."""


class ChildModel(BaseModel):

    a: int
    b: HumanDatetime


class ParentModel(BaseModel):
    """A parent model with human readable datetimes.

    All its properties and child model properties should have %Y-%m-%d %H:%M format."""

    c: ChildModel
    d: HumanDatetime


class ParentModelReport(ParentModel):
    """A ParentModel with high precision datetimes.

    All its properties and child model properties should have ISO format with nanoseconds."""

    model_config = ConfigDict(from_attributes=True)


# Create human readable instances
m1 = ChildModel(a=5, b="2023-01-06T10:00:20")
m2 = ParentModel(c=m1, d="2023-09-10T10:00")

# Create high precision parent model
m3 = ParentModelReport.model_validate(m2)  # <-- in fact, m2 is a database entry here,
#                                                but that's for tests

# Print models
print(
    f"""Human readable model
    python -> {m2.model_dump()}
    json   -> {m2.model_dump_json()}

High precision model
    python -> {m3.model_dump()}
    json   -> {m3.model_dump_json()}
"""
)
```

but I can't figure out how to set the `ParentModelReport` model to be precise for all datetime attributes.

Have some idea?

**Comment by adriangb** [comment]
Are you saying you can't/don't want to use the `HumanDatetime` annotation for each datetime attribute?

**Reply by etienne-monier** [reply]
I can, of course. It works very well to format datetimes in the `ParentModel` model. But how to modify the serializer for the `ParentModelReport` without re-writing everything?

**Reply by adriangb** [reply, marked as answer]
If you mean you want some way to "globally" serialize any `x: datetime` annotation in that model and all of it's children as if they were `x: HumanDatetime` then no, there is no way to do that, you'll have to replace all of them / rewrite everything.

**Reply by etienne-monier** [reply]
Ouch ! Saaaad :sob: 
Ok, I'll then write it twice.
Thanks for your help!

**Comment by ffdarkpenguin** [comment]
How about this in the base class?

```   
    @field_serializer('*', mode='plain', check_fields=False)
    def serialize_datetime(self, value):
        if isinstance(value, datetime):
            ret = value.isoformat(sep=' ', timespec='milliseconds')
            return ret
        return value```
